sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedProcessorBuilderImpl.java

```diff
@@ -209,6 +214,92 @@ public boolean isStarted() {
+
+                        String sessionTokenLsn = feedResponse.getSessionToken();
+                        String parsedSessionToken = sessionTokenLsn.substring(sessionTokenLsn.indexOf(pkRangeIdSeparator));
+                        String[] segments = parsedSessionToken.split(segmentSeparator);
+                        String latestLsn = segments[0];
+
+                        if (segments.length >= 2) {
+                            // default to Global LSN
+                            latestLsn = segments[1];
+                        }
+
+                        ChangeFeedProcessorState changeFeedProcessorState = new ChangeFeedProcessorState()
+                            .setId(lease.getId())
+                            .setHostName(lease.getOwner())
+                            .setLeaseToken(lease.getLeaseToken())
+                            .setLastUpdate(ZonedDateTime.parse(lease.getTimestamp()).toInstant());
+
+
+                        // An empty list of documents returned means that we are current (zero lag)
+                        if (feedResponse.getResults() == null || feedResponse.getResults().size() == 0) {
+                            changeFeedProcessorState.setEstimatedLag(0);
+                            changeFeedProcessorState.setContinuationToken(latestLsn);
+
+                            return changeFeedProcessorState;
+                        }
+
+                        changeFeedProcessorState.setContinuationToken(feedResponse.getResults().get(0).get(lsnPropertyName).asText(null));
+
+                        try {
+                            changeFeedProcessorState.setContinuationTokenTimestamp(Instant.ofEpochMilli(Long.valueOf(
+                                    feedResponse.getResults().get(0).get(tsPropertyName).asText("0"))));
+                        } catch (NumberFormatException ex) {
+                            logger.warn("Unexpected Cosmos _ts found", ex);
+                            changeFeedProcessorState.setContinuationTokenTimestamp(null);
+                        }
+
+                        Integer currentLsn = 0;
+                        Integer estimatedLag = 0;
+                        try {
+                            currentLsn = Integer.valueOf(feedResponse.getResults().get(0).get(lsnPropertyName).asText("0"));
+                            estimatedLag = Integer.valueOf(latestLsn);
+                            estimatedLag = estimatedLag - currentLsn + 1;
+                            changeFeedProcessorState.setEstimatedLag(estimatedLag);
+                        } catch (NumberFormatException ex) {
+                            logger.warn("Unexpected Cosmos LSN found", ex);
+                            changeFeedProcessorState.setEstimatedLag(-1);
+                        }
+
+                        return changeFeedProcessorState;
+                    });
+            })
+            .collectList()
+            .map(Collections::unmodifiableList);
```

Comment:
Again - feel free to ignore for consistency etc. - I am still struggling with Java not having a clear way to distinguish ReadOnlyList in the contract. But form my intuition I would expect it to be clearly documented in the Api doc comments if a method returns an unmodifiable list - so would it make sense to add a comment along tehse lines above? Also wouldn't it make sense to also return an unmodifiable list in the short-cut - when the list is empty) above to be consistent?

Reply:
yep, I've updated the doc to capture this is a read only list.

Reply:
also fixed the empty list early returned case (thanks for catching that up).